enhancements/installer/gcp-ipi-shared-vpc.md

```diff
@@ -0,0 +1,260 @@
+- Provisioning and/or configuring the Service Accounts for the Installer and
+  Cluster are to be done manually by the user prior to running the installer.
+
+## Proposal
+
+The installer should be able to install a GCP IPI cluster using Shared VPC
+networking. This requires some resources to be created and configured in
+advance, such as the network and subnets. Other resources can be created in
+advance when the necessary host project permissions are not desired, such as
+the firewall rules and/or DNS configuration. The installer, and some operators,
+will be updated to enable this functionality. They will become capable of
+creating and managing resources that exist in a different (Host) project from
+the cluster (Service Project). They will also become capable of consuming
+pre-existing firewall and/or DNS configurations.
+
+### Workflow Description
+
+Prior to executing the installer, the Shared VPC, Network, and Subnets will be
+manually created in the Host Project and configured for Shared VPC (XPN)
+configuration with the Service Project. A service account will also be created
+in the Service Project and granted sufficient permissions in the Host Project.
+When minimal permissions are desired in the Host Project, additional resources
+will be manually created and configured in the Host Project to include:
+firewall rules, api compute address, ingress compute address, and the api,
+api-int, and \*.apps DNS records.
+
+After all of these resources are created, they will be used in a manually
+created install-config.yaml to configure the cluster accordingly. At a minimum,
+the credentialsMode will be set to Passthrough, and the network,
+computeNetwork, controlPlaneNetwork, and networkProjectID will be configured to
+reference the Shared VPC network. When minimial permissions are desired in the
+Host Project, the additional parameters will be configured to enable the
+cluster to consume them as well.
+
+The installation process can then proceed as any other IPI workflow by running
+the create manifests, create ignition, and/or create cluster sub-commands.
+
+### API Extensions
+
+#### api.dns
+
+The current DNSZone struct does not appear to be able to store which project
+the zone is in. The ID variable can be used to store fully qualified path
+names, which will include the project. However, this may require modification
+of all consumers of this variable to be able to parse this information.
+Alternatively, a second variable can be added to store the projectID and the
+consumers can use that information when linking to the DNS Zone.
+
+#### api.ingress
+
+The current ingress structures do not appear to be able to store an address to
+be used when creating the ingress load balancer(s). New field(s) may be needed
+to enable this functionality. They can then be passed to the gcp lagacy
+provider for creating the ingress load balancers.
+
+#### installconfig.platform.gcp.networkProjectID
```

Comment:
So, something that comes to mind is that we do already have some customers using XPN, though via UPI IIRC. Reading through this enhancement, it seems that all the networking resources end up in the second project, is that the desired outcome here or is there a use case where some resource might want to be in the second project, and others in the main project? (It's been a while since I looked at XPN in depth)

Reply:
For GCP UPI XPN, everything that lands in the "host project" does so because it is a requirement for Shared VPC configuration. Anything that CAN exist in the Service Project will do so. This does not change for IPI... with the exception of the "public DNS zone", where we will actually be enabling it to exist in "any project."